### PR TITLE
Hoisting common expressions out of If blocks [retry]

### PR DESCRIPTION
**Pull request by Gamrix**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65645

Summary:
This is a retry of PR: https://github.com/pytorch/pytorch/pull/59492

Latest Changes: Added more tests, added the getOrCreateDB pattern, updated logic to remove unnecessary checks
addressed all comments.

Adding code to find common expressions from the two subblocks of an if
operation and hoist them before the if block.
This also allows Dead Code Elimination to
then eliminate some if blocks.

Test Plan:
python test_jit.py TestIfHoisting

Differential Revision: [D33302065](https://our.internmc.facebook.com/intern/diff/D33302065)

### CODE REVIEW

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302065).

File: torch/csrc/jit/passes/symbolic_shape_analysis.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include <torch/csrc/jit/ir/ir.h>
 #include <torch/csrc/jit/ir/ir_views.h>
 #include <torch/csrc/jit/jit_log.h>
+#include <torch/csrc/jit/passes/common_expression_hoisting.h>
```
Comment (eellison) [inline torch/csrc/jit/passes/symbolic_shape_analysis.cpp:9]: nit, dont need this here

**Review by eellison — APPROVED** [review]
stamp

**Comment by Gamrix** [comment]
@gamrix has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33302065).

**Pull request closed without merging**